Present the immersive model in the client app

jeanhbr opened the pull request:
#### 0d7a1ea90a00e80f609be99dee6e93c8251f9918
<pre>
Present the immersive model in the client app
<a href="https://bugs.webkit.org/show_bug.cgi?id=303541">https://bugs.webkit.org/show_bug.cgi?id=303541</a>
<a href="https://rdar.apple.com/165833224">rdar://165833224</a>

Reviewed by Etienne Segonzac.

Add WK delegate methods to enable the client to display the remote model inside their Immersive Space.
Plumb the context ID of the remote model up to the UI Process to create the remote environment view.

* LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt:
* LayoutTests/model-element/immersive/model-element-immersive-basic.html:
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline-expected.txt:
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline.html:
Updated test results.

* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::requestImmersive):
(WebCore::DocumentImmersive::exitImmersive):
Ensure the client has presented/dismissed the immersive environment before completed the immersive enter/exit request.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::allowImmersiveElement const):
(WebCore::ChromeClient::presentImmersiveElement const):
(WebCore::ChromeClient::dismissImmersiveElement const):
(WebCore::ChromeClient::canEnterImmersiveElement const): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _allowImmersiveElementFromURL:completion:]):
(-[WKWebView _presentImmersiveElement:completion:]):
(-[WKWebView _dismissImmersiveElement:]):
(-[WKWebView _canEnterImmersiveElementFromURL:completion:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKImmersiveEnvironmentDelegate.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::allowImmersiveElementFromURL const):
(WebKit::PageClientImplCocoa::presentImmersiveElement const):
(WebKit::PageClientImplCocoa::dismissImmersiveElement const):
(WebKit::PageClientImplCocoa::canEnterImmersiveElementFromURL): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::allowImmersiveElementFromURL const):
(WebKit::PageClient::presentImmersiveElement const):
(WebKit::PageClient::dismissImmersiveElement const):
(WebKit::PageClient::canEnterImmersiveElementFromURL): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::allowImmersiveElementFromURL const):
(WebKit::WebPageProxy::presentImmersiveElement const):
(WebKit::WebPageProxy::dismissImmersiveElement const):
(WebKit::WebPageProxy::canEnterImmersiveElementFromURL): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::allowImmersiveElement const):
(WebKit::WebChromeClient::presentImmersiveElement const):
(WebKit::WebChromeClient::dismissImmersiveElement const):
(WebKit::WebChromeClient::canEnterImmersiveElement const): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::allowImmersiveElement):
(WebKit::WebPage::presentImmersiveElement):
(WebKit::WebPage::dismissImmersiveElement):
(WebKit::WebPage::canEnterImmersiveElement): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView webView:allowImmersiveEnvironmentFromURL:completion:]):
(-[TestRunnerWKWebView webView:presentImmersiveEnvironment:completion:]):
(-[TestRunnerWKWebView webView:dismissImmersiveEnvironment:]):
(-[TestRunnerWKWebView webView:canPresentImmersiveEnvironmentFromURL:completion:]): Deleted.
Plumb the present and dismiss signals up to the UI Process.

Canonical link: <a href="https://commits.webkit.org/303987@main">https://commits.webkit.org/303987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0cd2253426eba1c11641ccc1a0244a4b4644f91e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6746 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45452 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141815 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136102 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7287 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6608 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102644 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137179 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/120317 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83437 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1629 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114144 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38454 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144460 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6416 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39035 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111025 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6499 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5400 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111274 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28217 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4816 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116588 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60200 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6468 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34800 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/6308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69934 "Built successfully") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6420 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2139,8 +2139,50 @@ - (void)_canEnterImmersiveElementFromURL:(const URL&)url completion:(CompletionH
+    auto completionBlock = makeBlockPtr(WTFMove(completion));
+    [immersiveEnvironmentDelegate webView:self presentImmersiveEnvironment:environmentView.get() completion:^(NSError * _Nullable error) {
+        if (error) {
+            completion(false);
```
completion is moved on line 2164!

etiennesegonzac replied:
I think it might be clearer ans less error prone to inline the `makeBlockPtr` like I see done here:
https://github.com/WebKit/WebKit/blob/a2e4e97e1a95de676ee82061f30bfd7e74dacf69/Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm#L689-L695

etiennesegonzac replied:
and maybe do the same across the file for consistency

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2139,8 +2139,50 @@ - (void)_canEnterImmersiveElementFromURL:(const URL&)url completion:(CompletionH
+        return;
+    }
+
+    auto environmentView = adoptNS([[UIView alloc] initWithFrame:CGRectZero]);
```
RetainPtr

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2139,8 +2139,50 @@ - (void)_canEnterImmersiveElementFromURL:(const URL&)url completion:(CompletionH
+    }
+
+    auto environmentView = adoptNS([[UIView alloc] initWithFrame:CGRectZero]);
+    auto remoteModelView = adoptNS([[_UIRemoteView alloc] initWithFrame:CGRectZero pid:[self _webProcessIdentifier] contextID:contextID.toUInt64()]);
```
same

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2139,8 +2139,50 @@ - (void)_canEnterImmersiveElementFromURL:(const URL&)url completion:(CompletionH
+    [remoteModelView setAutoresizingMask:(UIViewAutoresizingNone | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleBottomMargin)];
+
+    auto completionBlock = makeBlockPtr(WTFMove(completion));
+    [immersiveEnvironmentDelegate webView:self presentImmersiveEnvironment:environmentView.get() completion:^(NSError * _Nullable error) {
```
let's double check but I think we should pass `environmentView.autorelease()`

etiennesegonzac commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3064,7 +3065,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
-    void canEnterImmersiveElementFromURL(const URL&, CompletionHandler<void(bool)>&&);
+    void allowImmersiveElementFromURL(const URL&, CompletionHandler<void(bool)>&&);
+    void presentImmersiveElement(const WebCore::LayerHostingContextIdentifier, CompletionHandler<void(bool)>&&);
+    void dismissImmersiveElement(CompletionHandler<void()>&&);
```
should these be `const` too?

etiennesegonzac commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -1690,7 +1690,9 @@ class WebPage final : public API::ObjectImpl<API::Object::Type::BundlePage>, pub
-    void canEnterImmersiveElement(const WebCore::Element&, CompletionHandler<void(bool)>&&);
+    void allowImmersiveElement(const WebCore::Element&, CompletionHandler<void(bool)>&&);
+    void presentImmersiveElement(const WebCore::Element&, const WebCore::LayerHostingContextIdentifier, CompletionHandler<void(bool)>&&);
+    void dismissImmersiveElement(const WebCore::Element&, CompletionHandler<void()>&&);
```
same question

jeanhbr replied:
These functions are using `sendWithAsyncReply()` which I think cannot be used in `const` methods.

jeanhbr committed:
Present the immersive model in the client app
https://bugs.webkit.org/show_bug.cgi?id=303541
rdar://165833224

Reviewed by Etienne Segonzac.

Add WK delegate methods to enable the client to display the remote model inside their Immersive Space.
Plumb the context ID of the remote model up to the UI Process to create the remote environment view.

* LayoutTests/model-element/immersive/model-element-immersive-basic-expected.txt:
* LayoutTests/model-element/immersive/model-element-immersive-basic.html:
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline-expected.txt:
* LayoutTests/model-element/immersive/model-element-immersive-hidden-inline.html:
Updated test results.

* Source/WebCore/dom/DocumentImmersive.cpp:
(WebCore::DocumentImmersive::requestImmersive):
(WebCore::DocumentImmersive::exitImmersive):
Ensure the client has presented/dismissed the immersive environment before completed the immersive enter/exit request.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::allowImmersiveElement const):
(WebCore::ChromeClient::presentImmersiveElement const):
(WebCore::ChromeClient::dismissImmersiveElement const):
(WebCore::ChromeClient::canEnterImmersiveElement const): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _allowImmersiveElementFromURL:completion:]):
(-[WKWebView _presentImmersiveElement:completion:]):
(-[WKWebView _dismissImmersiveElement:]):
(-[WKWebView _canEnterImmersiveElementFromURL:completion:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKImmersiveEnvironmentDelegate.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::allowImmersiveElementFromURL const):
(WebKit::PageClientImplCocoa::presentImmersiveElement const):
(WebKit::PageClientImplCocoa::dismissImmersiveElement const):
(WebKit::PageClientImplCocoa::canEnterImmersiveElementFromURL): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::allowImmersiveElementFromURL const):
(WebKit::PageClient::presentImmersiveElement const):
(WebKit::PageClient::dismissImmersiveElement const):
(WebKit::PageClient::canEnterImmersiveElementFromURL): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::allowImmersiveElementFromURL const):
(WebKit::WebPageProxy::presentImmersiveElement const):
(WebKit::WebPageProxy::dismissImmersiveElement const):
(WebKit::WebPageProxy::canEnterImmersiveElementFromURL): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::allowImmersiveElement const):
(WebKit::WebChromeClient::presentImmersiveElement const):
(WebKit::WebChromeClient::dismissImmersiveElement const):
(WebKit::WebChromeClient::canEnterImmersiveElement const): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::allowImmersiveElement):
(WebKit::WebPage::presentImmersiveElement):
(WebKit::WebPage::dismissImmersiveElement):
(WebKit::WebPage::canEnterImmersiveElement): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView webView:allowImmersiveEnvironmentFromURL:completion:]):
(-[TestRunnerWKWebView webView:presentImmersiveEnvironment:completion:]):
(-[TestRunnerWKWebView webView:dismissImmersiveEnvironment:]):
(-[TestRunnerWKWebView webView:canPresentImmersiveEnvironmentFromURL:completion:]): Deleted.
Plumb the present and dismiss signals up to the UI Process.

Canonical link: https://commits.webkit.org/303987@main
